source/SoundEmojiSynthesizer.cpp

```diff
@@ -207,6 +207,10 @@ ManagedBuffer SoundEmojiSynthesizer::pull()
                 done = true;
                 if (renderComplete || status & EMOJI_SYNTHESIZER_STATUS_STOPPING)
                 {
+                    // Wait for a hanful of milliseconds while we finish playing the last buffer we sent out ...
+                    fiber_sleep( buffer.length() / sampleRate );
```

Comment:
At worst the buffer length could be 1 byte short of a full buffer but as we generally sample at comparitively high rates, this should result in only a few milliseconds at most.
I can instrument this up and record it, but I'm not concerned as the buffers are usually quite small (<1k).

I haven't thrown this on a scope yet, as running this with a MakeCode program requires a custom stack, which I do not have access to right now.

Reply:
The main reason I ask is because I'm not familiar with what kind of numbers we are looking into.
What is the smallest, average and larger buffer size and sample rate? Based on that, what would be the min, average and max values we would expect here?
Does the buffer size usually match the sample rate? double it, halve it?

Is the sampleRate value clamped somewhere? Are we protected from a division by zero?

What happens if the calculation results in a `fiber_sleep()`?


>I haven't thrown this on a scope yet, as running this with a MakeCode program requires a custom stack, which I do not have access to right now.

Might be worth seeing how easy it is to replicate the original report with a CODAL-only programme and then see how this PR affects it.

Reply:
> What happens if the calculation results in a fiber_sleep(0)?

Thats an easy one - if you attempt to `fiber_sleep()` with a `0` or negative number, you'll effectively just get yourself descheduled for 1 scheduler loop and then you'll be back on the ready queue.

Reply:
But your point about sample rate is valid - it is certainly possible to hand-craft a stream with zero-samples-per-sec set, although quite why that would be useful I don't know right now.

We do later also do this: https://github.com/lancaster-university/codal-microbit-v2/blob/90756513c3ff795a80eb5ef54aa6abe67d3d0a6e/source/SoundEmojiSynthesizer.cpp#L233 which would have the same issue...

The only place we set the sample rate is here: https://github.com/lancaster-university/codal-microbit-v2/blob/90756513c3ff795a80eb5ef54aa6abe67d3d0a6e/source/SoundEmojiSynthesizer.cpp#L328 and we cna introduce a check here for sub-zero values. (I'll patch that in this PR)

Reply:
Looking at this, I have two questions. 

1) Carlos and I have just tracked through the numbers and it looks from visual inspection like we're sleeping for 512/44100 milliseconds, which I don't think fiber_sleep can handle well. Do we have a unit mismatch between sample rate and fiber_sleep?

2) If (as we guess) this should be (512/(44100/1000)) then we're still going to be calling fiber_sleep(11.6099773243) which will end up as 12ms sleep requested. With the jitter involved from the scheduler, isn't there a risk here that we end up sleeping both longer and shorter than that, which will stop us smoothly chaining sounds together?

It feels like it would be better to wait here until we get an event from the audio subsystem that the transfer has completed? @finneyj I know you were specifically optimising this so that buffers could be smoothly chained, so perhaps I've not understood the special casing here? WDYT?